src/acsets/ColumnImplementations.jl

```diff
@@ -0,0 +1,592 @@
+end
+
+
+"""
+A column for a vec-backed injective hom
+"""
+struct DenseInjectiveHom <: Column{Int,Int}
+  m::SentinelVecMap{Int, DefaultVal{Int, 0}}
+  i::InjectiveIC{Int, Int, SentinelVecMap{Int, DefaultVal{Int,0}}}
+end
+
+# Compatibility hack to support unique_index
+function Columns.preimage(dom, c::DenseInjectiveHom, y::Int, ::InconsistentPreimageFlag)
+  # @warn "using old convention for preimage of unique-indexed homomorphisms"
+  c.i.inverse[y]
+end
+
+function Columns.preimage_multi(dom, c::DenseInjectiveHom, ys, ::InconsistentPreimageFlag)
+  # @warn "using old convention for preimage of unique-indexed homomorphisms"
+  c.i.inverse[ys]
+end
+
+
+"""
+A column for a dict-backed unindexed hom with key type K
+"""
+struct SparseHom{K} <: Column{K,K}
+  m::PartialDictMap{K, K}
+  i::TrivialIC{K, K}
+end
+
+"""
+A column for a dict-backed indexed hom with key type K
+"""
+struct SparseIndexedHom{K} <: Column{K,K}
+  m::PartialDictMap{K, K}
+  i::StoredPreimageIC{K, K, Set{K},
+                      DefaultDictMap{K, Set{K}, DefaultEmpty{Set{K}}}}
+end
+
+"""
+A column for a dict-backed injective hom with key type K
+"""
+struct SparseInjectiveHom{K} <: Column{K,K}
+  m::PartialDictMap{K,K}
+  i::InjectiveIC{K, K, PartialDictMap{K,K}}
+end
+
+"""
+A column for a vec-backed unindexed attr
+"""
+struct DenseAttr{T} <: Column{Int,T}
+  m::PartialVecMap{T}
+  i::TrivialIC{Int, T}
+end
+
+"""
+A column for a vec-backed indexed attr
+"""
+struct DenseIndexedAttr{T} <: Column{Int,T}
+  m::PartialVecMap{T}
+  i::StoredPreimageIC{Int, T, SortedSet{Int},
+                      DefaultDictMap{T, SortedSet{Int}, DefaultEmpty{SortedSet{Int}}}}
+end
+
+"""
+A column for a vec-backed unindexed attr
+"""
+struct DenseInjectiveAttr{T} <: Column{Int,T}
+  m::PartialVecMap{T}
+  i::InjectiveIC{Int, T, PartialDictMap{T, Int}}
+end
+
+# Compatibility hack
+function Columns.preimage(dom, c::DenseInjectiveAttr{T}, y::T, ::InconsistentPreimageFlag) where {T}
+  # @warn "using old convention for preimage of unique-indexed attributes"
+  get(c.i.inverse.d, y, 0)
+end
+
+function Columns.preimage_multi(dom, c::DenseInjectiveAttr, ys, ::InconsistentPreimageFlag)
+  # @warn "using old convention for preimage of unique-indexed attributes"
+  broadcast(ys) do y
+    get(c.i.inverse.d, y, 0)
+  end
+end
+
+"""
+A column for a dict-backed unindexed hom with key type K
+"""
+struct SparseAttr{K, T} <: Column{K,T}
+  m::PartialDictMap{K, T}
+  i::TrivialIC{K, T}
+end
+
+"""
+A column for a dict-backed indexed hom with key type K
+"""
+struct SparseIndexedAttr{K, T} <: Column{K,T}
+  m::PartialDictMap{K, T}
+  i::StoredPreimageIC{K, T, Set{K},
+                      DefaultDictMap{T, Set{K}, DefaultEmpty{Set{K}}}}
+end
+
+"""
+A column for a dict-backed injective hom with key type K
+"""
+struct SparseInjectiveAttr{K, T} <: Column{K,T}
+  m::PartialDictMap{K,T}
+  i::InjectiveIC{K, T, PartialDictMap{T, K}}
+end
+
+
+@data HomOrAttr begin
```

Comment:
This whole bit for selecting the default column types is slick.

Reply:
Thanks!